[GeoMechanicsApplication] Reuse the nodal extrapolator in U-Pw small strain element

avdg81 opened the pull request:
**📝 Description**
Replace the hard-coded extrapolation matrices for U-Pw small strain elements by requesting the nodal extrapolator to calculate the extrapolation matrices. Note that at present, the nodal extrapolator supports triangles and quadrilaterals only. And since the member functions of the U-Pw small strain element support linear elements only, this PR considers linear triangles and linear quadrilaterals only.

Furthermore, existing integration tests have been extended to cover the changes made.

**🆕 Changelog**
- Member functions `CalculateExtrapolationMatrix` of class `UPwSmallStrainElement` for linear triangles and linear quadrilaterals no longer return hard-coded matrices. Instead, they now use a linear nodal extrapolator to carry out their task.
- Existing test cases in file `test_elements.py` for linear elements (triangles, quadrilaterals, tetrahedra, and hexahedra) now also check the vertical stress component at the bottom nodes. For tetrahedra and hexahedra, it covers untouched code.
- All test cases in file `test_elements.py` now retrieve the displacement field at the end time from a GiD output file rather than from the analysis database. Unfortunately, I couldn't extend that approach to the stress and strain tensors, since writing such output to GiD output files is not supported for neither 15-noded triangles nor 10-noded tetrahedra.
- Cleaned up the lists of requested output items. We now only ask for the output items that we actually check. This will help with code coverage analysis, since previously code appeared to be covered whereas it wasn't. Most output items were actually calculated (which results in line hits when performing code coverage analysis), but the computed values were simply not checked. Therefore, we only ask for `NODAL_CAUCHY_STRESS_TENSOR` when the test involves linear elements.
- GiD output files have to be text files in order to process them with our test tools. Where previously binary output files were requested, we now ask for ASCII files.

avdg81 committed:
Reuse the linear nodal extrapolator for linear U-Pw triangles

- Extended the test for 2D linear triangles such that it checks the
  vertical stress component at the bottom nodes. This check uses the
  nodal stress tensors at the end time, which are read from the GiD
  output file (in ASCII format).
- Replaced the hard-coded extrapolation matrix for 2D linear triangles
  by requesting the linear nodal extrapolator to calculate the
  extrapolation matrix for the element's geometry and integration
  scheme.

avdg81 committed:
Reuse the linear nodal extrapolator for linear U-Pw quadrilaterals

- Extended the test for 2D linear quadrilaterals such that it checks the
  vertical stress component at the bottom nodes. This check uses the
  nodal stress tensors at the end time, which are read from the GiD
  output file (in ASCII format).
- Replaced the hard-coded extrapolation matrix for 2D linear
  quadrilaterals by requesting the linear nodal extrapolator to
  calculate the extrapolation matrix for the element's geometry and
  integration scheme.
- Made the linear nodal extrapolator instances `const`.

avdg81 committed:
Changed the GidPostMode to GiD_PostAscii for the other element tests

This is to prepare for extending the other element tests with checks for
the vertical stress component at the bottom nodes.

avdg81 committed:
Check that for all element tests the GiD output file exists

Also define and pass the bottom node IDs for each test, even though they
are not being used yet.

avdg81 committed:
For all linear element tests, assert the vertical stress at the bottom

avdg81 committed:
Removed unused parameter of `assert_linear_elastic_block`

The vertical stress component at the bottom nodes won't be asserted by
this method. As a result, parameter `bottom_node_ids` had become
redundant, and therefore it has been removed.

avdg81 committed:
No longer request for output that is not being asserted

avdg81 committed:
Fetch the displacements from the GiD output file

Also, we no longer explicitly assert that the output file exists. That
is done implicitly now.

avdg81 committed:
Give an expected value a descriptive name

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1542,43 +1543,35 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateExtrapolationMatrix(Bounde
 template <>
 void UPwSmallStrainElement<2, 3>::CalculateExtrapolationMatrix(BoundedMatrix<double, 3, 3>& rExtrapolationMatrix)
 {
```
I was wondering is it better to move the content to the template instead of these two instantiations?  Right now there is only difference between them in TNumNodes value. The template can check TNumNodes values to avoid using it for other values. What is your opinion?

mnabideltares replied:
I think template is redundant here as no argument is passed to template. Can the template be removed from here?

avdg81 replied:
Nice suggestion. I have moved the code of the template specializations to the template itself, using `if constexpr` to distinguish the supported cases. Since this is a pure refactoring, the function's behavior should not change. Therefore, the tested conditions reflect the earlier template specializations. To make the code easier to read, I have also extracted two local helper functions that return extrapolation matrices for linear tetrahedra and linear hexahedra.

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1542,43 +1543,35 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateExtrapolationMatrix(Bounde
-    // node. Rows: nodes Columns: GP
-
     // Triangle_2d_3
     // GI_GAUSS_2
```
If the template will be used, then these comments are obsolete. If not, I would suggest to use constexpr for dimension instead //Triangle_2d_3

avdg81 replied:
Removed.

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1542,43 +1543,35 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateExtrapolationMatrix(Bounde
 void UPwSmallStrainElement<2, 4>::CalculateExtrapolationMatrix(BoundedMatrix<double, 4, 4>& rExtrapolationMatrix)
 {
     // Quadrilateral_2d_4
     // GI_GAUSS_2
```
As above

mnabideltares replied:
Same as above, in the case can be without template ...

avdg81 replied:
Removed.

markelov208 commented on applications/GeoMechanicsApplication/tests/test_quad_4n.gid/ProjectParameters.json:
```diff
@@ -90,8 +90,8 @@
                         "plane_output":        [],
-                        "nodal_results":       ["DISPLACEMENT","TOTAL_DISPLACEMENT","ROTATION","WATER_PRESSURE","REACTION","REACTION_WATER_PRESSURE","POINT_LOAD","LINE_LOAD","SURFACE_LOAD","NORMAL_CONTACT_STRESS","TANGENTIAL_CONTACT_STRESS","NORMAL_FLUID_FLUX","VOLUME_ACCELERATION","NODAL_CAUCHY_STRESS_TENSOR","NODAL_DAMAGE_VARIABLE","NODAL_JOINT_WIDTH","NODAL_JOINT_DAMAGE"],
-                        "gauss_point_results": ["MOMENT","FORCE","GREEN_LAGRANGE_STRAIN_TENSOR","CAUCHY_STRESS_TENSOR","TOTAL_STRESS_TENSOR","VON_MISES_STRESS","FLUID_FLUX_VECTOR","PERMEABILITY_MATRIX","DAMAGE_VARIABLE","JOINT_WIDTH","LOCAL_STRESS_VECTOR","LOCAL_RELATIVE_DISPLACEMENT_VECTOR","LOCAL_FLUID_FLUX_VECTOR","LOCAL_PERMEABILITY_MATRIX"]
+                        "nodal_results":       ["DISPLACEMENT","NODAL_CAUCHY_STRESS_TENSOR"],
```
Thanks for cleaning the lists.

avdg81 replied:
You're welcome :-)

markelov208 reviewed (commented):
Hi Anne, thank you for the very nice unification and cleaning work.

mnabideltares commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1542,43 +1543,35 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateExtrapolationMatrix(Bounde
+        << "Extrapolation matrix has unexpected number of columns: " << result.size1()
+        << " (expected 3)" << std::endl;
+    rExtrapolationMatrix = result;
 }
```
The number of rows and colums are checked here. However, they are hard programmed to be allocated and filled correctly. So, I think checking a correctly allocated arrays is just extra work and time for the code and increases the line in the code. What do you think about it?

avdg81 replied:
I have added these checks, since the "extrapolator" returns a matrix with dynamic sizes. In other words, its sizes are unknown at compile time. However, the very next action is to assign this variable-sized matrix to a matrix with fixed sizes, which are known at compile time (i.e. `Matrix` vs. `bounded_matrix<...>`). To avoid that we attempt to assign a matrix that has mismatching sizes, we explicitly check the variable-sized matrix.

mnabideltares commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1542,43 +1543,35 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateExtrapolationMatrix(Bounde
+    KRATOS_ERROR_IF_NOT(result.size2() == 4)
+        << "Extrapolation matrix has unexpected number of columns: " << result.size1()
+        << " (expected 4)" << std::endl;
+    rExtrapolationMatrix = result;
```
Same as above, is check necessary?

avdg81 replied:
Yes, I would strongly recommend to do it. See my explanation above.

mnabideltares commented on applications/GeoMechanicsApplication/tests/test_quad_4n.gid/ProjectParameters.json:
```diff
@@ -80,7 +80,7 @@
                             "WriteDeformedMeshFlag": "WriteUndeformed",
                             "WriteConditionsFlag":   "WriteElementsOnly",
-                            "GiDPostMode":           "GiD_PostBinary",
+                            "GiDPostMode":           "GiD_PostAscii",
```
Is there a reason to change binary to ascii? As long as we don't need to look at the values inside, binary is preffered, as it leads to compacter file.

avdg81 replied:
Helper class `GidOutputFileReader` can only handle GiD output files in text format. Since we now use that helper class in our tests (for checking the displacement fields), we must use textual output files rather than binary ones.

mnabideltares commented on applications/GeoMechanicsApplication/tests/test_quad_8n.gid/ProjectParameters.json:
```diff
@@ -80,7 +80,7 @@
                             "WriteDeformedMeshFlag": "WriteUndeformed",
                             "WriteConditionsFlag":   "WriteElementsOnly",
-                            "GiDPostMode":           "GiD_PostBinary",
+                            "GiDPostMode":           "GiD_PostAscii",
```
Same as above

avdg81 replied:
See my previous reply.

mnabideltares commented on applications/GeoMechanicsApplication/tests/test_tetra_10n.gid/ProjectParameters.json:
```diff
@@ -80,7 +80,7 @@
                             "WriteConditionsFlag":   "WriteElementsOnly",
-                            "GiDPostMode":           "GiD_PostBinary",
+                            "GiDPostMode":           "GiD_PostAscii",
                             "MultiFileFlag":         "SingleFile"
```
Same as avove

avdg81 replied:
See my previous reply.

mnabideltares commented on applications/GeoMechanicsApplication/tests/test_tetra_4n.gid/ProjectParameters.json:
```diff
@@ -80,7 +80,7 @@
                             "WriteDeformedMeshFlag": "WriteUndeformed",
                             "WriteConditionsFlag":   "WriteElementsOnly",
-                            "GiDPostMode":           "GiD_PostBinary",
+                            "GiDPostMode":           "GiD_PostAscii",
```
Same as above, do we need asccii?

avdg81 replied:
See my previous reply.

mnabideltares reviewed (commented):
@avdg81 Thank you for refactoring and cleaning up the code.

avdg81 committed:
Processed Gennady's review suggestions

- Extracted two local helper functions that return the extrapolation
  matrices for linear tetrahedra and linear hexahedra.
- No longer use template specializations to handle the supported cases
  for calculating the extrapolation matrices, but use `if constexpr`
  instead.
- The code for linear triangles and linear quadrilaterals is now being
  shared.
- Fixed a minor mistake in the error message when the number of columns
  does not match our expectation. It used to report the number of rows
  rather than the number of columns. That has been corrected.

markelov208 reviewed (approved):
Hi Anne, good to be merged.

mnabideltares reviewed (approved):
Looks good to me
